src/autopas/CMakeLists.txt

```
@@ -12,6 +12,22 @@ file(
     "*.cuh"
 )
 
+# If rules-based tuning is disabled, we must remove all files which link to antlr. As this includes generated files,
+# where we can't easily change the file, we disable the files here.
+if (NOT AUTOPAS_ENABLE_RULES_BASED_TUNING)
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/parser_generated/autopas_generated_rule_syntax/RuleLanguageBaseVisitor.h")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/parser_generated/autopas_generated_rule_syntax/RuleLanguageBaseVisitor.cpp")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/parser_generated/autopas_generated_rule_syntax/RuleLanguageVisitor.h")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/parser_generated/autopas_generated_rule_syntax/RuleLanguageVisitor.cpp")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/parser_generated/autopas_generated_rule_syntax/RuleLanguageLexer.h")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/parser_generated/autopas_generated_rule_syntax/RuleLanguageLexer.cpp")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/parser_generated/autopas_generated_rule_syntax/RuleLanguageParser.h")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/parser_generated/autopas_generated_rule_syntax/RuleLanguageParser.cpp")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/RuleBasedProgramParser.h")
+    list(REMOVE_ITEM MY_SRC "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/RuleBasedProgramParser.cpp")
+endif ()
+
+
```

If I may suggest a slightly different approach, I would exclude the whole rules base tuning directory:

```suggestion
if (NOT AUTOPAS_ENABLE_RULES_BASED_TUNING)
    file(GLOB_RECURSE RULES_FILES "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/*.cpp"
            "${AUTOPAS_SOURCE_DIR}/src/autopas/tuning/tuningStrategy/ruleBasedTuning/*.h")
    # Loop over each file in RULES_FILES and remove it from MY_SRC
    foreach(FILE ${RULES_FILES})
        list(REMOVE_ITEM MY_SRC ${FILE})
    endforeach()
endif ()
```

Then you would not need all the `#ifdef` in `RuleBasedTuning.h/.cpp`, but only one `#ifdef` in `TuningStrategyFactory.cpp`, sth like:
```cpp
    case TuningStrategyOption::ruleBasedTuning: {
#ifdef AUTOPAS_ENABLE_RULES_BASED_TUNING
      tuningStrategy = std::make_unique<RuleBasedTuning>(searchSpace, /*verify mode*/ false, info.ruleFileName);
#else
utils::ExceptionHandler::exception(
    "Cannot use the TuningStrategy ruleBasedTuning without AUTOPAS_ENABLE_RULES_BASED_TUNING=OFF.");
#endif
      break;
    }
```

---

I don't have a strong opinion either way. The advantage of not excluding `RuleBasedTuning.h/.cpp` is that the interface still exists. I vaguely remember @FG-TUM suggesting not doing this - I would be interested in his view on this.

---

I'm a bit confused. I thought I already wrote something somewhere about this section and topic...

1. I'd handle the CMake code way simpler in a way that is more resistant to project structure changes:
```suggestion
if (NOT AUTOPAS_ENABLE_RULES_BASED_TUNING)
    list(FILTER MY_SRC EXCLUDE REGEX ".*parser-generated.*")
endif ()
```

2. I prefer the `ifdef` in the `RuleBasedTuning` class for several reasons:
  - You don't have to consider every include wherever in the project it happens or will happen. When you write any code you should not have to think about "do I have to treat this include special?". It should resolve this by itself. This might also affect the tests.
  - The `ifdef` is where it actually should affect the code.
  - Consistency with how we already do it for things like e.g. specialized loggers.